### ISSUE TITLE
Expose schema of sync webhooks response 

### ISSUE THREAD

**Issue by lkostrowski** [saleor/saleor#11905]
### What I'm trying to achieve
I want to statically type @saleor/app-sdk to ensure endpoints return proper structure for Sync Events like taxes or payments

I don't want to re-declare (hardcode) schema on the client, to avoid de-sync and double source of truth.

### Describe a proposed solution
Prepare a schema that can be consumed by the SDK and other clients.

It can use format that can be transpiled into popular formats, like d.ts, but also Swift or Java

It can be [json-schema](https://json-schema.org/), but maybe graphQL can work as well

Transpilation can work both on Saleor Code codebase CICD (artifacts for n languages) or it can be just a file, that clients transpile.
Probably best DX will be if popular languages are supported without additional step.

File should be marked with `CODEOWNERS` pointing to every Saleor team to ensure clients are aware

File can be placed at Core repository, but can be also a separate "contract repository"

### Other solutions I've tried and won't work
I am hardcoding values

**Comment by krzysztofwolski** [comment]
Would love if the OpenAPI would be chosen:
- Docs could generate automatically
- could generate types for the SDK
- a lot of tools that could be used to test both the core implementation and consumers

For the file location: not a fan of the current location of gql schema, not quite clear when to find it when browsing code the first time. `/schema` directory for gql and sync hooks would IMO improved (but require a migration period for all the old scripts to update the path)

**Comment by mmiszy** [comment]
I'm convinced we should use JSON schema. As a matter of fact, that's what we've started using in saleor-app-payment template:

https://github.com/saleor/saleor-app-payment-template/tree/canary/src/schemas

- JSON schemas
- A script that compiles them to .d.ts files (using `json-schema-to-typescript`)
- A script that compiles them to .mjs files (using `ajv/dist/standalone`)

**Comment by lkostrowski** [comment]
Dropping idea from @patrys 

Instead of building artifacts inside workflow and committing them, build them and upload to Github storage. They should be tracked per release and connected to the release, like assets here: https://github.com/saleor/saleor/releases/tag/3.12.2

All schemas should be clearly versioned

**Comment by andrzejewsky** [comment]
+1 to `json-schema` this is actually the most common approach and as was mentioned it's already been used in our projects.

**One point**: transformation to other languages/syntax - Please test them before any wider usage, at least for our common cases, I remember that I had some problems with it. It was years ago so probably it has significantly changed, but worth checking.

**Comment by lkostrowski** [comment]
> Before we introduced the subscription webhooks I was checking the possibility to add the schema for the webhooks with pydantic. Finally, we moved with the subscription webhooks as it was a lot of more benefits than adding typing to old payloads. But it sounds like a perfect thing for this RFC. Personally, I would keep this with core, as we could use pydantic models to handle the response from App, and generate the schema based on the models. Below you can find a sample for one webhook.
> 
> ```python
> from __future__ import annotations
> 
> from typing import Any, Dict, List, Optional
> 
> from enum import Enum
> from pydantic import BaseModel, Field
> from decimal import Decimal
> 
> class Result(str, Enum):
>     CHARGE_SUCCESS = 'CHARGE_SUCCESS'
>     CHARGE_FAILURE = 'CHARGE_FAILURE'
> 
> 
> class TransactionInitializeSession(BaseModel):
>     pspReference: str
>     time: Optional[str]
>     externalUrl: Optional[str]
>     data: Dict[Any, Any] = Field(None)
>     message: Optional[str] = Field(None, description="Message related to the action.")
>     result: Result = Field(..., description="The final result of the transaction intialize webhook.")
>     amount: Decimal = Field(..., gte=Decimal(0))
> 
> 
> 
> print(TransactionInitializeSession.schema_json(indent=2))
> ```
> 
> As a respone we will get:
> 
> ```json
> {
>   "title": "TransactionInitializeSession",
>   "type": "object",
>   "properties": {
>     "pspReference": {
>       "title": "Pspreference",
>       "type": "string"
>     },
>     "time": {
>       "title": "Time",
>       "type": "string"
>     },
>     "externalUrl": {
>       "title": "Externalurl",
>       "type": "string"
>     },
>     "data": {
>       "title": "Data",
>       "type": "object"
>     },
>     "message": {
>       "title": "Message",
>       "description": "Message related to the action.",
>       "type": "string"
>     },
>     "result": {
>       "description": "The final result of the transaction intialize webhook.",
>       "allOf": [
>         {
>           "$ref": "#/definitions/Result"
>         }
>       ]
>     },
>     "amount": {
>       "title": "Amount",
>       "gte": 0,
>       "type": "number"
>     }
>   },
>   "required": [
>     "pspReference",
>     "result",
>     "amount"
>   ],
>   "definitions": {
>     "Result": {
>       "title": "Result",
>       "description": "An enumeration.",
>       "enum": [
>         "CHARGE_SUCCESS",
>         "CHARGE_FAILURE"
>       ],
>       "type": "string"
>     }
>   }
> }
> ```
> 
> Accodring to pydantic docs:
> 
> > The generated schemas are compliant with the specifications: [JSON Schema Core](https://json-schema.org/latest/json-schema-core.html), [JSON Schema Validation](https://json-schema.org/latest/json-schema-validation.html) and [OpenAPI](https://github.com/OAI/OpenAPI-Specification).

@korycins do you suggest to build json-schema from pydantic? 

This is important decision, because json-schema is not a source anymore. Im afraid that it will cause some problems.

My proposal is to have these scenarios:
1. pydantic loads json-schema to validate (extra rules can be added on top of that, as a 2nd line of validation)
2. JS/TS code can load json-schema to tools like `ajv` to achieve the same in js/ts
3. compiles can generate json-schema -> `*.d.ts`

If we start with pydantic classes, we will have:
1. pydantic -> json-schema -> ajv
2. pydantic ->json-schema -> .d.ts

json-schema as an artifact no longer can be controlled "by hand", so if there is need to any manual improvement (syntax, description), we can be limited (unless pydantic to json-schema is super well done and covers everything)

I can't prove these problems will appear, its a general feeling that we will have 2 layers of auto-generation and its sound dangerous

**Comment by korycins** [comment]
@lkostrowski I suggest adopting the same approach for building a JSON schema as we have for building a GraphQL schema. We already have a data class within Saleor's code that stores the details of the response that a webhook can return. Therefore, it should be feasible to convert these classes to Pydantic models and provide a command or endpoint to share them with you.

Personally, I believe it is crucial to ensure that the generated JSON schema is always related to the response models defined on the Saleor side. This way, we can guarantee that the schema is a valid contract. I am not certain if Pydantic can be customized to meet our specific requirements. Although I tested it with some old async payloads a year or two ago and had no issues, we will likely need to try it with two or three sync webhooks to confirm that it can handle all of the improvements we would like to add.

**Comment by przlada** [comment]
I agree with @korycins to start with a test. It seems that Pydantic provides ways to configure and customize JSON schema generated from its models: https://docs.pydantic.dev/usage/schema/

**Comment by lkostrowski** [comment]
@mmiszy @przlada please handover this topic from Appstore to DEX team

**Comment by mmiszy** [comment]
In @saleor/dex we don't have any competencies related to the core and/or Python. What are the exact requirements on our side once this issue is handed to us? @przlada @lkostrowski

**Comment by lkostrowski** [comment]
I don't think there is anything related to core. Its purely about tooling and DX. AFAIK @przlada already created feature that builds json-schema from Python validators/model. 

To finish this, we need to take json-schema and attach additional tooling, like conversion to .d.ts, attaching it to the Saleor release artifacts etc

@przlada can pass the details

**Comment by mmiszy** [comment]
So the outstanding work will be in the CLI, correct?

**Comment by lkostrowski** [comment]
@mmiszy I think the MVP should be to attach json-schema + .d.ts to each Saleor release artifact.

Then we can connect it to app-templates and SDK to fetch it, just like we fetch schema.
If we see any CLI usages, we can add it later
